File: android/app/src/main/java/com/porttennanttandoori/pos/data/repository/OrdersRepository.kt
package com.porttennanttandoori.pos.data.repository

import com.porttennanttandoori.pos.data.model.OrderDetailDto
import com.porttennanttandoori.pos.data.model.OrderEvent
import com.porttennanttandoori.pos.data.model.OrderListItemDto
import com.porttennanttandoori.pos.data.model.OrderListPageDto
import com.porttennanttandoori.pos.data.model.OrderStatus
import com.porttennanttandoori.pos.data.model.OrderStatusDefinitionDto
import com.porttennanttandoori.pos.data.model.PaymentStatusDefinitionDto
import com.porttennanttandoori.pos.data.model.UpdateOrderStatusRequest
import com.porttennanttandoori.pos.data.network.ApiService
import com.porttennanttandoori.pos.data.network.OrderEventsClient
import java.io.IOException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine

/** Holds the terminal's working set of orders in memory, kept current by a REST refresh plus
 * whatever OrderListenerService feeds it from the SSE stream. No local Room cache yet - orders
 * are re-fetched from the API on cold start, which is fine as long as the terminal has network. */
class OrdersRepository(
    private val apiService: ApiService,
    private val eventsClient: OrderEventsClient,
) {
    private val _orders = MutableStateFlow<List<OrderListItemDto>>(emptyList())
    val orders: StateFlow<List<OrderListItemDto>> = _orders.asStateFlow()

    // Statuses are per-restaurant and admin-configurable (Admin/StatusDefinitionsController),
    // not a fixed enum - fetched once and kept ordered by DisplayOrder so nextStatus() can walk
    // the same sequence the admin Configurations page and web Orders stepper use.
    private val _orderStatusDefinitions = MutableStateFlow<List<OrderStatusDefinitionDto>>(emptyList())
    val orderStatusDefinitions: StateFlow<List<OrderStatusDefinitionDto>> = _orderStatusDefinitions.asStateFlow()

    private val _paymentStatusDefinitions = MutableStateFlow<List<PaymentStatusDefinitionDto>>(emptyList())
    val paymentStatusDefinitions: StateFlow<List<PaymentStatusDefinitionDto>> = _paymentStatusDefinitions.asStateFlow()

    /** Orders still sitting in the restaurant's starting status (IsDefault=true on the status
     * definition, "Pending" out of the box) - these are the ones NewOrderAlarm keeps ringing for
     * until staff confirms or cancels them. */
    val ordersAwaitingConfirmation: Flow<List<OrderListItemDto>> =
        combine(orders, orderStatusDefinitions) { currentOrders, definitions ->
            val startingStatus = definitions.firstOrNull { it.isDefault }?.name ?: return@combine emptyList()
            currentOrders.filter { it.status == startingStatus }
        }

    fun eventStream() = eventsClient.stream()

    suspend fun loadStatusDefinitions() {
        try {
            val orderStatusesResponse = apiService.listOrderStatusDefinitions()
            if (orderStatusesResponse.isSuccessful) {
                _orderStatusDefinitions.value = orderStatusesResponse.body()?.data.orEmpty()
            }
            val paymentStatusesResponse = apiService.listPaymentStatusDefinitions()
            if (paymentStatusesResponse.isSuccessful) {
                _paymentStatusDefinitions.value = paymentStatusesResponse.body()?.data.orEmpty()
            }
        } catch (e: IOException) {
            // Non-fatal: the status dropdown just stays empty until this succeeds on a later
            // refresh; the order list itself doesn't depend on these.
        } catch (e: kotlinx.serialization.SerializationException) {
            // Same as above - a response-shape mismatch here shouldn't crash the terminal.
        }
    }

    suspend fun refresh(): Result<Unit> {
        return try {
            val response = apiService.listOrders()
            if (response.isSuccessful) {
                _orders.value = response.body()?.data?.items.orEmpty()
                Result.success(Unit)
            } else {
                Result.failure(IOException("Failed to load orders (${response.code()})."))
            }
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: kotlinx.serialization.SerializationException) {
            Result.failure(e)
        }
    }

    /** Order History's search/filter/pagination - a plain one-shot fetch, not folded into the
     * live `orders` StateFlow above (that one stays SSE-driven and unpaginated by design, since
     * the New Orders tab needs its full working set in memory, not a page of it). */
    suspend fun searchOrders(
        status: String?,
        search: String?,
        dateFrom: String?,
        dateTo: String?,
        page: Int,
        pageSize: Int,
    ): Result<OrderListPageDto> {
        return try {
            val response = apiService.searchOrders(status, search, dateFrom, dateTo, page, pageSize)
            val pageResult = response.body()?.data
            if (response.isSuccessful && pageResult != null) {
                Result.success(pageResult)
            } else {
                Result.failure(IOException("Failed to search orders (${response.code()})."))
            }
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: kotlinx.serialization.SerializationException) {
            Result.failure(e)
        }
    }

    suspend fun getOrder(orderId: String): Result<OrderDetailDto> {
        return try {
            val response = apiService.getOrder(orderId)
            val detail = response.body()?.data
            if (response.isSuccessful && detail != null) {
                Result.success(detail)
            } else {
                Result.failure(IOException("Failed to load order (${response.code()})."))
            }
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: kotlinx.serialization.SerializationException) {
            Result.failure(e)
        }
    }

    suspend fun updateStatus(orderId: String, status: OrderStatus, note: String? = null): Result<OrderDetailDto> {
        return try {
            val response = apiService.updateOrderStatus(orderId, UpdateOrderStatusRequest(status, note))
            val detail = response.body()?.data
            if (response.isSuccessful && detail != null) {
                applyStatusLocally(orderId, status)
                Result.success(detail)
            } else {
                Result.failure(IOException("Failed to update order status (${response.code()})."))
            }
        } catch (e: IOException) {
            Result.failure(e)
        } catch (e: kotlinx.serialization.SerializationException) {
            Result.failure(e)
        }
    }

    /** Called from OrderListenerService when an SSE event arrives, so the list updates without
     * waiting for the next full refresh(). */
    suspend fun onEvent(event: OrderEvent) {
        when (event) {
            is OrderEvent.OrderCreated -> refresh()
            is OrderEvent.OrderStatusChanged -> applyStatusLocally(event.orderId, event.status)
            is OrderEvent.PaymentReceived -> refresh()
        }
    }

    private fun applyStatusLocally(orderId: String, status: OrderStatus) {
        _orders.value = _orders.value.map { if (it.id == orderId) it.copy(status = status) else it }
    }
}

File: android/lib/api/orders_repository.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import 'api_client.dart';
import 'models.dart';
import 'sse_client.dart';

/// Holds the terminal's working set of orders in memory, kept current by a
/// REST refresh plus whatever the SSE stream feeds it. No local cache - orders
/// are re-fetched from the API on cold start. Direct port of
/// OrdersRepository.kt's semantics onto a StreamController-based API.
class OrdersRepository {
  OrdersRepository({required this.apiClient, required this.sseClient});

  final ApiClient apiClient;
  final SseClient sseClient;

  final _ordersController = StreamController<List<OrderListItem>>.broadcast();
  final _orderStatusDefsController = StreamController<List<OrderStatusDefinition>>.broadcast();
  final _paymentStatusDefsController = StreamController<List<PaymentStatusDefinition>>.broadcast();

  List<OrderListItem> _orders = [];
  List<OrderStatusDefinition> _orderStatusDefs = [];
  List<PaymentStatusDefinition> _paymentStatusDefs = [];

  List<OrderListItem> get orders => _orders;
  List<OrderStatusDefinition> get orderStatusDefinitions => _orderStatusDefs;
  List<PaymentStatusDefinition> get paymentStatusDefinitions => _paymentStatusDefs;

  Stream<List<OrderListItem>> get ordersStream => _ordersController.stream;
  Stream<List<OrderStatusDefinition>> get orderStatusDefinitionsStream => _orderStatusDefsController.stream;
  Stream<List<PaymentStatusDefinition>> get paymentStatusDefinitionsStream => _paymentStatusDefsController.stream;

  /// Orders still sitting in the restaurant's starting status
  /// (isDefault=true, "Pending" out of the box) - these are the ones the
  /// new-order alarm/takeover keeps ringing for until staff confirm/cancel.
  List<OrderListItem> get ordersAwaitingConfirmation {
    final startingStatus = _orderStatusDefs.where((d) => d.isDefault).firstOrNull?.name;
    if (startingStatus == null) return const [];
    return _orders.where((o) => o.status == startingStatus).toList();
  }

  Future<void> loadStatusDefinitions() async {
    try {
      _orderStatusDefs = await apiClient.listOrderStatusDefinitions();
      _orderStatusDefsController.add(_orderStatusDefs);
      debugPrint('[Orders] status defs loaded: ${_orderStatusDefs.map((d) => '${d.name}(default=${d.isDefault})').join(', ')}');
    } catch (e) {
      debugPrint('[Orders] loadStatusDefinitions (order) failed: $e');
    }
    try {
      _paymentStatusDefs = await apiClient.listPaymentStatusDefinitions();
      _paymentStatusDefsController.add(_paymentStatusDefs);
    } catch (e) {
      debugPrint('[Orders] loadStatusDefinitions (payment) failed: $e');
    }
  }

  Future<void> refresh() async {
    // loadStatusDefinitions() only otherwise runs once at startup - retry it
    // on every refresh (every OrderCreated/PaymentReceived event, not just
    // cold start) until it actually succeeds, or the new-order alarm and
    // takeover screen silently stop matching anything forever.
    if (_orderStatusDefs.isEmpty || _paymentStatusDefs.isEmpty) {
      await loadStatusDefinitions();
    }
    final page = await apiClient.listOrders(historyOnly: false);
    _orders = page.items;
    _ordersController.add(_orders);
    debugPrint('[Orders] refreshed: ${_orders.map((o) => '#${o.orderNumber}:${o.status}').join(', ')}');
    debugPrint('[Orders] awaiting confirmation: ${ordersAwaitingConfirmation.length}');
  }

  Future<OrderListPage> searchOrders({
    String? status,
    String? paymentStatus,
    String? search,
    String? dateFrom,
    String? dateTo,
    bool? historyOnly,
    required int page,
    required int pageSize,
  }) =>
      apiClient.searchOrders(
        status: status,
        paymentStatus: paymentStatus,
        search: search,
        dateFrom: dateFrom,
        dateTo: dateTo,
        historyOnly: historyOnly,
        page: page,
        pageSize: pageSize,
      );

  Future<OrderDetail> getOrder(String id) => apiClient.getOrder(id);

  Future<OrderDetail> updateStatus(String orderId, String status, {String? note}) async {
    final detail = await apiClient.updateOrderStatus(orderId, status, note: note);
    _applyStatusLocally(orderId, status);
    return detail;
  }

  Future<OrderDetail> updatePaymentStatus(String orderId, String paymentStatus) async {
    final detail = await apiClient.updatePaymentStatus(orderId, paymentStatus);
    _applyPaymentStatusLocally(orderId, paymentStatus);
    return detail;
  }

  Stream<OrderEvent> eventStream({String? bearerToken}) => sseClient.stream(bearerToken: bearerToken);

  Future<void> onEvent(OrderEvent event) async {
    switch (event) {
      case OrderCreatedEvent():
        await refresh();
      case OrderStatusChangedEvent(:final orderId, :final status):
        _applyStatusLocally(orderId, status);
      case PaymentReceivedEvent():
        await refresh();
    }
  }

  // The cached `_orders` list backs the New Orders tab only (history is always
  // fetched fresh via searchOrders) - once a status change lands an order on
  // a terminal status (Completed/Cancelled) it belongs in History now, so
  // drop it here rather than waiting for the next full refresh().
  void _applyStatusLocally(String orderId, String status) {
    if (_isTerminalStatus(status)) {
      _orders = _orders.where((o) => o.id != orderId).toList();
    } else {
      _orders = _orders.map((o) => o.id == orderId ? o.copyWith(status: status) : o).toList();
    }
    _ordersController.add(_orders);
  }

  bool _isTerminalStatus(String status) {
    final countsAsCompleted = _orderStatusDefs.where((d) => d.name == status).firstOrNull?.countsAsCompleted ?? false;
    return countsAsCompleted || status == cancelledStatusName(_orderStatusDefs);
  }

  void _applyPaymentStatusLocally(String orderId, String paymentStatus) {
    _orders = _orders.map((o) => o.id == orderId ? o.copyWith(paymentStatus: paymentStatus) : o).toList();
    _ordersController.add(_orders);
  }

  void dispose() {
    _ordersController.close();
    _orderStatusDefsController.close();
    _paymentStatusDefsController.close();
  }
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}

/// Mirrors StatusColors.kt's nextStatusAfter: one step past the current
/// status's displayOrder, falling back to itself if there's nowhere to go.
String nextStatusAfter(String current, List<OrderStatusDefinition> definitions) {
  final sorted = [...definitions]..sort((a, b) => a.displayOrder.compareTo(b.displayOrder));
  final index = sorted.indexWhere((d) => d.name == current);
  if (index == -1 || index + 1 >= sorted.length) return current;
  return sorted[index + 1].name;
}

/// The terminal status - matched by name like the admin dashboard - falling
/// back to the literal name if unconfigured.
String cancelledStatusName(List<OrderStatusDefinition> definitions) =>
    definitions.where((d) => d.name.toLowerCase() == 'cancelled').firstOrNull?.name ?? 'Cancelled';
